File: mycodes/155-MinStack.java
//49% Time
//93% Space

class MinStack {

    Stack<Integer> stack, minStack;
    
    /** initialize your data structure here. */
    public MinStack() {
        stack = new Stack<>();
        minStack = new Stack<>();
    }
    
    public void push(int x) {
        stack.push(x);
        if (minStack.isEmpty() || getMin() >= x) minStack.push(x);
    }
    
    public void pop() {
        if (stack.isEmpty()) return;
        int elem = stack.pop();
        if (elem == getMin()) minStack.pop();
    }
    
    public int top() {
        return stack.peek();
    }
    
    public int getMin() {
        return minStack.peek();
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(x);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */

File: mycodes/155-MinStack.go
package main

type MinStack struct {
	Mins  []int
	Stack []int
}

func Constructor() MinStack {
	return MinStack{
		Mins:  make([]int, 0),
		Stack: make([]int, 0),
	}
}

func (this *MinStack) Push(val int) {
	this.Stack = append(this.Stack, val)
	if len(this.Mins) == 0 || this.Mins[len(this.Mins)-1] >= val {
		this.Mins = append(this.Mins, val)
	}
}

func (this *MinStack) Pop() {
	val := this.Stack[len(this.Stack)-1]
	this.Stack = this.Stack[:len(this.Stack)-1]
	if val == this.GetMin() {
		this.Mins = this.Mins[:len(this.Mins)-1]
	}
}

func (this *MinStack) Top() int {
	return this.Stack[len(this.Stack)-1]
}

func (this *MinStack) GetMin() int {
	return this.Mins[len(this.Mins)-1]
}

/**
 * Your MinStack object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Push(val);
 * obj.Pop();
 * param_3 := obj.Top();
 * param_4 := obj.GetMin();
 */
